### ISSUE TITLE
No speedup on CUDA

### ISSUE THREAD

**Issue by idontgetoutmuch** [AccelerateHS/accelerate#376]
I am writing this as a ticket because I find google groups make
following threads difficult.

The following code is supposed to generate 384 batches ("paths") of
10,000 random numbers ("a path") in parallel, 384 being the number of
cores on my GPU.

```
./.stack-work/install/x86_64-osx/lts-8.0/8.0.2/bin/nvidia-device-query 
CUDA device query (Driver API, statically linked)
CUDA driver version 7.5
Detected 1 CUDA capable device

Device 0: GeForce GT 650M
  CUDA capability:                    3.0
  CUDA cores:                         384 cores in 2 multiprocessors (192 cores/MP)
```

```haskell
{-# LANGUAGE TypeOperators #-}

{-# OPTIONS_GHC -Wall #-}

module Main ( main ) where

import Data.Array.Accelerate              as A
import Data.Array.Accelerate.LLVM.Native  as CPU
import Data.Array.Accelerate.LLVM.PTX     as GPU
import Prelude hiding ((^))
import qualified Prelude                  as P

bsd :: Exp (Int, Double) -> Exp (Int, Double)
bsd n = let
  m = (A.fst n * 1103515245 + 12345) `mod` 2^(lift (31 :: Int))
  y = A.fromIntegral (A.fst n)
  x = y / 2147483648.0
  in
  lift $
  (m, x)

main :: IO ()
main = mapM_ putStrLn $
       P.map show $
       P.map (P.sum . toList . CPU.run . fold (+) 0 . A.map A.snd . rngsFull)
             (P.take 10 [1, 385..])

rngsFull :: Int -> Acc (Array (DIM1 :. Int) (Int, Double))
rngsFull p = A.scanl (\s x -> bsd (lift (A.fst x + A.fst s, A.snd s)))
                     (lift (1 :: Int, 0.5 :: Double))
                     (seedPair p 384 10000)

seeds :: Int -> Int -> Int -> Acc (Array DIM2 Int)
seeds p m n = transpose $ use $
              A.fromList (Z :. n :. m)
                         (([p..m+p-1]) P.++ (P.replicate (m * (n - 1)) 0))

seedPair :: Int -> Int -> Int -> Acc (Array DIM2 (Int, Double))
seedPair p m n = A.zip xs ys
  where
    xs = seeds p m n
    ys = use $ A.fromList (Z :. m :. n) (P.replicate (m * n) 0.0)
```

When I run this on the CPU I get

```
 $ time ./Gillespie +RTS -N1 > out.txt
time ./Gillespie +RTS -N1 > out.txt
Gillespie: forkOS_entry: interrupted
Gillespie: forkOS_entry: interrupted

real	0m5.195s
user	0m5.315s
sys	0m0.305s
 $ time ./Gillespie +RTS -N2 > out.txt
time ./Gillespie +RTS -N2 > out.txt
Gillespie: forkOS_entry: interrupted
Gillespie: forkOS_entry: interrupted

real	0m3.698s
user	0m5.583s
sys	0m0.553s
 $ time ./Gillespie +RTS -N4 > out.txt
time ./Gillespie +RTS -N4 > out.txt
Gillespie: Gillespie: forkOS_entry: interruptedforkOS_entry: interrupted


real	0m2.903s
user	0m6.798s
sys	0m0.977s
```

so we see some parallelism (NB using +RTS -s gave some confusing
information which I won't reproduce here).

But using the GPU where I think I should see something like a x384
speedup I get

```
 $ time ./Gillespie > out.txt
time ./Gillespie > out.txt

real	0m4.450s
user	0m3.773s
sys	0m0.315s
```

**Comment by tmcdonell** [comment]
There are a couple things going on here, so I'll try to unpack it. Feel free to ask for clarification etc. on each point.

 * Each GPU core is individually much weaker than a CPU core. They derive their speed from parallelism. This comes a few forms;
    - When an instruction is stalled (e.g waiting on a memory read), the core swaps out a different thread to continue execution (a warp, actually). This is in contrast to a CPU, which relies on speculative execution, branch prediction, etc. to hide stalls. In particular, although your device has 384 physical cores, you'll need many more threads than that to actually saturate the arithmetic units / memory bandwidth.
    - This difference in architecture is what allows GPUs to scale; Titan Xp has 3584 cores for example. I'd expect your GT 650 to be roughly the same power as your cpu, actually, but it depends on the workload (quite possibly we have the same machine).

* Since Accelerate is (for the moment) a runtime compiler, your benchmark is actually mostly measuring compilation time. If you build accelerate and friends with debugging on you can see this:

```sh
$ ./issue376-cpu +ACC -ddump-phases -ddump-exec -ACC +RTS -N -RTS
[   2.318] phase sharing-recovery: 482.701 ms (wall), 2.184 s (cpu)
[   2.318] phase rewrite-segment-offset: 482.960 ms (wall), 2.184 s (cpu)
[   2.321] phase array-fusion: 490.809 ms (wall), 2.194 s (cpu)
[   2.399] phase compile: 73.323 ms (wall), 77.878 ms (cpu)
[   2.857] exec: scanS 63.226 ms (wall), 456.956 ms (cpu), 7.23 x speedup
[   2.867] exec: fold 1.656 ms (wall), 9.938 ms (cpu), 6.00 x speedup
[   2.867] phase execute: 65.339 ms (wall), 467.349 ms (cpu), 7.15 x speedup
1920659.9522743225
...

$ ./issue376-ptx +ACC -ddump-phases -ddump-exec -ACC
[   0.433] phase sharing-recovery: 301.485 ms (wall), 301.405 ms (cpu)
[   0.433] phase rewrite-segment-offset: 301.767 ms (wall), 301.687 ms (cpu)
[   0.434] phase array-fusion: 302.846 ms (wall), 302.762 ms (cpu)
[   0.721] phase compile: 253.823 ms (wall), 254.397 ms (cpu)
[   0.953] exec: scan <<< 4, 992, 24304 >>> 50.000 µs (wall), 49.000 µs (cpu), 79.796 ms (gpu)
[   1.029] exec: fold <<< 4, 1024, 12544 >>> 22.000 µs (wall), 32.000 µs (cpu), 35.622 ms (gpu)
[   1.029] phase execute: 116.638 ms (wall), 307.647 ms (cpu)
3972646.614979394
```

I rewrote your program to use `run1`, which is usually Super Important™:

```haskell
{-# LANGUAGE RebindableSyntax #-}
{-# LANGUAGE TypeOperators    #-}
{-# OPTIONS_GHC -Wall #-}

module Main ( main ) where

import Data.Array.Accelerate              as A
import Data.Array.Accelerate.LLVM.Native  as CPU
import Data.Array.Accelerate.LLVM.PTX     as GPU
import Prelude                            hiding ((^), (==))
import qualified Prelude                  as P

bsd :: Exp (Int, Double) -> Exp (Int, Double)
bsd n = let
  m = (A.fst n * 1103515245 + 12345) `mod` 2^(lift (31 :: Int))
  y = A.fromIntegral (A.fst n)
  x = y / 2147483648.0
  in
  lift $
  (m, x)

main :: IO ()
main
  = mapM_ putStrLn
  $ P.map (show . flip indexArray Z)
  $ P.map (go . singleton)
  $ P.take 10 [1, 385..]
  where
    go          = GPU.run1 (A.sum . A.map A.snd . rngsFull)
    singleton x = fromList Z [x]

rngsFull :: Acc (Scalar Int) -> Acc (Array DIM2 (Int,Double))
rngsFull p
  = A.scanl (\s x -> bsd (lift (A.fst x + A.fst s, A.snd s)))
            (constant (1, 0.5))
            (seedPair (constant (Z:.384:.10000)) p)

seedPair :: Exp DIM2 -> Acc (Scalar Int) -> Acc (Array DIM2 (Int,Double))
seedPair sh p
  = A.zip (seeds sh p) (A.fill sh 0)

seeds :: Exp DIM2 -> Acc (Scalar Int) -> Acc (Array DIM2 Int)
seeds sh p
  = A.generate sh
  $ \ix -> let Z :. m :. n = unlift ix :: Z :. Exp Int :. Exp Int
           in  if n == 0
                 then m + the p
                 else 0
```

(also I wasn't sure why you were doing that sum on the host, so I just moved that to accelerate land as well).

The trace looks much nicer now:

```sh
$ ./issue376-ptx +ACC -ddump-phases -ddump-exec -ACC +RTS -N -RTS
[   0.154] phase sharing-recovery: 5.171 ms (wall), 19.579 ms (cpu)
[   0.155] phase rewrite-segment-offset: 5.641 ms (wall), 21.025 ms (cpu)
[   0.158] phase array-fusion: 7.006 ms (wall), 23.824 ms (cpu)
[   0.489] phase compile: 259.101 ms (wall), 330.922 ms (cpu)
[   0.776] exec: scan <<< 4, 992, 24304 >>> 47.000 µs (wall), 45.000 µs (cpu), 60.242 ms (gpu)
[   0.931] exec: foldAllM1 <<< 4, 1024, 12544 >>> 22.000 µs (wall), 20.000 µs (cpu), 36.394 ms (gpu)
[   0.938] phase execute: 98.024 ms (wall), 446.521 ms (cpu)
[   0.938] exec: foldAllM2 <<< 4, 1024, 12544 >>> 19.000 µs (wall), 29.000 µs (cpu), 93.728 µs (gpu)
3971731.962091446
[   0.938] exec: foldAllM2 <<< 1, 1024, 12544 >>> 18.000 µs (wall), 16.000 µs (cpu), 47.616 µs (gpu)
[   1.257] exec: scan <<< 4, 992, 24304 >>> 34.000 µs (wall), 33.000 µs (cpu), 63.244 ms (gpu)
[   1.397] exec: foldAllM1 <<< 4, 1024, 12544 >>> 18.000 µs (wall), 18.000 µs (cpu), 36.395 ms (gpu)
[   1.403] exec: foldAllM2 <<< 4, 1024, 12544 >>> 18.000 µs (wall), 17.000 µs (cpu), 95.168 µs (gpu)
[   1.407] phase execute: 102.143 ms (wall), 468.356 ms (cpu)
3963692.4895572662
[   1.407] exec: foldAllM2 <<< 1, 1024, 12544 >>> 17.000 µs (wall), 17.000 µs (cpu), 48.096 µs (gpu)
[   1.711] exec: scan <<< 4, 992, 24304 >>> 46.000 µs (wall), 44.000 µs (cpu), 61.212 ms (gpu)
[   1.854] exec: foldAllM1 <<< 4, 1024, 12544 >>> 20.000 µs (wall), 19.000 µs (cpu), 33.966 ms (gpu)
[   1.859] phase execute: 112.551 ms (wall), 451.310 ms (cpu)
[   1.859] exec: foldAllM2 <<< 4, 1024, 12544 >>> 23.000 µs (wall), 22.000 µs (cpu), 49.088 µs (gpu)
[   1.859] exec: foldAllM2 <<< 1, 1024, 12544 >>> 23.000 µs (wall), 23.000 µs (cpu), 27.584 µs (gpu)
3975233.0170230865
[   2.012] exec: scan <<< 4, 992, 24304 >>> 30.000 µs (wall), 29.000 µs (cpu), 29.594 ms (gpu)
[   2.091] phase execute: 50.311 ms (wall), 231.403 ms (cpu)
[   2.089] exec: foldAllM1 <<< 4, 1024, 12544 >>> 20.000 µs (wall), 19.000 µs (cpu), 17.739 ms (gpu)
[   2.091] exec: foldAllM2 <<< 4, 1024, 12544 >>> 20.000 µs (wall), 18.000 µs (cpu), 48.416 µs (gpu)
3975913.544488907
[   2.091] exec: foldAllM2 <<< 1, 1024, 12544 >>> 24.000 µs (wall), 23.000 µs (cpu), 26.432 µs (gpu)
[   2.229] exec: scan <<< 4, 992, 24304 >>> 30.000 µs (wall), 28.000 µs (cpu), 28.349 ms (gpu)
[   2.315] phase execute: 49.495 ms (wall), 222.912 ms (cpu)
[   2.311] exec: foldAllM1 <<< 4, 1024, 12544 >>> 18.000 µs (wall), 16.000 µs (cpu), 18.690 ms (gpu)
[   2.315] exec: foldAllM2 <<< 4, 1024, 12544 >>> 24.000 µs (wall), 22.000 µs (cpu), 25.664 µs (gpu)
[   2.315] exec: foldAllM2 <<< 1, 1024, 12544 >>> 18.000 µs (wall), 16.000 µs (cpu), 8.192 µs (gpu)
3968544.071954727
[   2.448] exec: scan <<< 4, 992, 24304 >>> 32.000 µs (wall), 30.000 µs (cpu), 28.057 ms (gpu)
[   2.511] exec: foldAllM1 <<< 4, 1024, 12544 >>> 18.000 µs (wall), 17.000 µs (cpu), 16.831 ms (gpu)
[   2.518] phase execute: 45.548 ms (wall), 202.248 ms (cpu)
[   2.518] exec: foldAllM2 <<< 4, 1024, 12544 >>> 24.000 µs (wall), 23.000 µs (cpu), 49.600 µs (gpu)
[   2.518] exec: foldAllM2 <<< 1, 1024, 12544 >>> 17.000 µs (wall), 16.000 µs (cpu), 31.072 µs (gpu)
3970245.5994205475
[   2.653] exec: scan <<< 4, 992, 24304 >>> 32.000 µs (wall), 30.000 µs (cpu), 28.508 ms (gpu)
[   2.724] exec: foldAllM1 <<< 4, 1024, 12544 >>> 20.000 µs (wall), 18.000 µs (cpu), 16.802 ms (gpu)
[   2.733] phase execute: 47.276 ms (wall), 214.880 ms (cpu)
[   2.733] exec: foldAllM2 <<< 4, 1024, 12544 >>> 46.000 µs (wall), 62.000 µs (cpu), 46.400 µs (gpu)
[   2.734] exec: foldAllM2 <<< 1, 1024, 12544 >>> 19.000 µs (wall), 18.000 µs (cpu), 8.480 µs (gpu)
3970222.126886368
[   2.860] exec: scan <<< 4, 992, 24304 >>> 32.000 µs (wall), 31.000 µs (cpu), 29.124 ms (gpu)
[   2.942] phase execute: 45.198 ms (wall), 208.287 ms (cpu)
[   2.942] exec: foldAllM2 <<< 1, 1024, 12544 >>> 24.000 µs (wall), 186.000 µs (cpu), 7.776 µs (gpu)
[   2.942] exec: foldAllM2 <<< 4, 1024, 12544 >>> 21.000 µs (wall), 393.000 µs (cpu), 28.832 µs (gpu)
[   2.936] exec: foldAllM1 <<< 4, 1024, 12544 >>> 34.000 µs (wall), 46.000 µs (cpu), 15.478 ms (gpu)
3968068.654352188
[   3.077] exec: scan <<< 4, 992, 24304 >>> 33.000 µs (wall), 30.000 µs (cpu), 27.108 ms (gpu)
[   3.146] phase execute: 45.264 ms (wall), 202.498 ms (cpu)
[   3.137] exec: foldAllM1 <<< 4, 1024, 12544 >>> 27.000 µs (wall), 25.000 µs (cpu), 16.172 ms (gpu)
[   3.147] exec: foldAllM2 <<< 4, 1024, 12544 >>> 20.000 µs (wall), 19.000 µs (cpu), 42.880 µs (gpu)
[   3.147] exec: foldAllM2 <<< 1, 1024, 12544 >>> 20.000 µs (wall), 18.000 µs (cpu), 25.728 µs (gpu)
3975511.1818180084
[   3.265] exec: scan <<< 4, 992, 24304 >>> 58.000 µs (wall), 56.000 µs (cpu), 27.083 ms (gpu)
[   3.314] exec: foldAllM1 <<< 4, 1024, 12544 >>> 78.000 µs (wall), 140.000 µs (cpu), 15.114 ms (gpu)
[   3.340] phase execute: 42.721 ms (wall), 192.389 ms (cpu)
3973543.7092838287
[   3.340] exec: foldAllM2 <<< 4, 1024, 12544 >>> 34.000 µs (wall), 31.000 µs (cpu), 47.104 µs (gpu)
```

or...

```sh
$ time ./issue376-cpu +RTS -N1 -RTS
...
        1.03 real         0.86 user         0.15 sys

$ time ./issue376-cpu +RTS -N8 -RTS
...
        0.50 real         1.60 user         0.31 sys

$ time ./issue376-ptx +RTS -N8 -RTS
...
        1.02 real         0.88 user         0.19 sys
```

Hope that clears up a bit what is going on.

**Comment by idontgetoutmuch** [comment]
Thanks very much for such a comprehensive reply. Maybe it's because I've just woken up but the ptx version still takes longer than the cpu version with -N8 and is the same for -N1. I was hoping for a x100 speedup.

Am I right to understand that a GPU core is about x100 *less* powerful than a CPU core so that even though I have 384 cores that's not going to give an overall improvement? So if I want a speedup then I need something with far more GPU cores?

I can rent an Amazon machine with 36 CPU cores (not sure how they compare to my macbook cores but roughly equivalent) and without going to e.g. Wuxi or Oak Ridge I am unlikely to be able to access more. I wonder what GPU device I would need to beat 36 CPU cores? And how much would it cost?

**Comment by tmcdonell** [comment]
It depends on your application.

GPUs have lots of memory bandwidth available, which CPUs can only match when the working set is small and fits in cache. This is what we're seeing here. Once the data grows a bit larger than cache, performance on the CPU will drop significantly. This is one reason GPUs are particularly suited to applications where there is lots of data to process (graphics, image processing, machine learning...).

There is also a question of price/performance ratio. An [E7-8867 v4](http://ark.intel.com/products/93804/Intel-Xeon-Processor-E7-8867-v4-45M-Cache-2_40-GHz) (18 cores) has an RRP of $4600, whereas the [Titan Xp](https://www.nvidia.com/en-us/geforce/products/10series/titan-xp/) is $1200. Each of these have their own strengths and weaknesses.

So yeah, your MBP may have a CPU and GPU of roughly equivalent power (under certain situations), but a different way to look at that might be that now you can use _both_ the CPU and the GPU at the same time, with the same Accelerate code.

**Comment by tmcdonell** [comment]
FWIW, Intel went through essentially the same problem you have now with the Xeon Phi. Anyway, I guess the question is answered now, so closing this. Feel free to reopen / start a new ticket if you still have questions.

**Comment by idontgetoutmuch** [comment]
Thanks very much for all your help. I don't follow what you mean by Intel having the same problem with Xeon Phi.

At the moment, I'd like the full paths not just the sum of the observations of the path (the sum was to avoid printing lots of output and force the computation) maybe `scanl` is killing the parallelism? I'll carry on playing and see what I can do.
